### PR TITLE
Transaction signer, callable SmartContracts, Update user connector.

### PR DESCRIPTION
**Pull request by digitaldonkey**

### CODE REVIEW

**Comment by digitaldonkey** [comment]
@amateescu Do I use the services wisely now? I'm sometimes a little unsure. Ii would be very happy to get your feedback on that.

File: ethereum_smartcontract/src/Entity/SmartContract.php [hunk 1]
```diff
@@ -38,43 +42,183 @@
...
+    return $deployed;
+  }
+
+  /**
+   * Contract Network Info Table.
+   *
+   * @return array
+   *    Table render array.
+   */
+  public function getDeployedAsTable() {
```
Comment (amateescu) [inline ethereum_smartcontract/src/Entity/SmartContract.php]: An entity object shouldn't provide methods that output render arrays, that's the job of a list builder or a specialized controller.

File: ethereum_smartcontract/src/SmartContractInterface.php [hunk 2]
```diff
@@ -2,10 +2,70 @@
...
 * Provides an interface defining an Example entity.
 */
 interface SmartContractInterface extends ConfigEntityInterface {
-// Add get/set methods for your configuration properties here.
+
+
+  /**
+   * @inheritdoc
+   */
+  public function status();
```
Comment (amateescu) [inline ethereum_smartcontract/src/SmartContractInterface.php]: This method is already defined by `ConfigEntityInterface` so we don't need to declare it again.

File: ethereum_txsigner/src/Entity/Txsigner.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  * @var string
+  */
+  protected $label;
+
+  /**
+   * Is active.
+   *
+   * @var boolean
+   */
+  protected $status;
```
Comment (amateescu) [inline ethereum_txsigner/src/Entity/Txsigner.php:50]: This is already defined by `ConfigEntityBase`, no need to redefine it here.

File: src/Controller/EthereumController.php [hunk 4]
```diff
@@ -21,15 +23,20 @@ class EthereumController extends ControllerBase {
    *
    * @var \Ethereum\Ethereum
    */
-  protected $client;
+  public $client;
```
Comment (amateescu) [inline src/Controller/EthereumController.php]: Why do we need this change?

File: src/Controller/EthereumController.php [hunk 5]
```diff
@@ -21,15 +23,20 @@ class EthereumController extends ControllerBase {
...
   private $debug = TRUE;
 
   /**
    * Constructs a new EthereumController.
+   *
+   * @param \Ethereum\Ethereum $web3
    */
-  public function __construct(Ethereum $ethereum_client) {
-    $this->client = $ethereum_client;
+  public function __construct(Ethereum $web3 = NULL) {
```
Comment (amateescu) [inline src/Controller/EthereumController.php]: This change should not be needed since we're are injecting the ethereum client from the container.

I assume that some other code is not initializing this class properly..

File: src/Controller/EthereumController.php [hunk 6]
```diff
@@ -92,9 +105,11 @@ public static function getServerOptionsArray($filter_enabled = FALSE){
    * Returns Ethereum Servers.
    *
    * @param bool $filter_enabled
-   *   (optional) Restrict to enabled servers. Defaults to FALSE.
+   *    Restrict to enabled servers. Defaults to FALSE (optional).
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
    *
-   * @return \Drupal\ethereum\EthereumServerInterface[]
+   * @return array of \Drupal\ethereum\Entity\EthereumServer
```
Comment (amateescu) [inline src/Controller/EthereumController.php]: What's the reason for this change? The previous code was using the standard notation for an array of objects.

File: src/Controller/EthereumController.php [hunk 7]
```diff
@@ -92,9 +105,11 @@ public static function getServerOptionsArray($filter_enabled = FALSE){
    * Returns Ethereum Servers.
    *
    * @param bool $filter_enabled
-   *   (optional) Restrict to enabled servers. Defaults to FALSE.
+   *    Restrict to enabled servers. Defaults to FALSE (optional).
```
Comment (amateescu) [inline src/Controller/EthereumController.php]: Drupal coding standards say that (optional) needs to come first, like it was before :)

File: src/Controller/EthereumController.php [hunk 8]
```diff
@@ -108,6 +123,26 @@ public static function getServers($filter_enabled = FALSE) {
...
+   * @return EthereumServerInterface
+   *   Ethereum current Default server config entity.
+   *
+   * @throws \Exception
+   */
+  public static function getDefaultServer() {
+    $current = \Drupal::config('ethereum.settings')->get('current_server');
+    $server = \Drupal::entityTypeManager()->getStorage('ethereum_server')->load($current);
+    if (!$server) {
+      throw new \Exception('Current default server (' . $current . ')does not exist.');
```
Comment (amateescu) [inline src/Controller/EthereumController.php]: Missing a space between `)` and `does`.

File: src/Entity/EthereumServer.php [hunk 9]
```diff
@@ -105,13 +107,17 @@ public function isDefaultServer() {
...
+    $return = [
+      'error' => FALSE,
+      'message' => 'Server ' . $this->id . ' (' . $this->url .') is up and on listening on Ethereum network id ' . $this->network_id
+    ];
     try {
-      /** @var \Ethereum\Ethereum $client */
-      $client = \Drupal::service('ethereum.client_factory')->get($this->getUrl());
+      /** @var \Ethereum\Ethereum $web3 */
+       $web3 = new Ethereum($this->url);
+      // @todo How to load the \Drupal::service('ethereum.client') with a updated url?
```
Comment (amateescu) [inline src/Entity/EthereumServer.php]: The previous code was already handling that..

File: src/Entity/EthereumServer.php [hunk 10]
```diff
@@ -129,8 +135,53 @@ public function validateConnection() {
...
     return $return;
   }
 
+  /**
+   * Server info as render Array Table.
+   *
+   * @return array
+   *    Table render array.
+   */
+  public function getServerInfoAsTable() {
```
Comment (amateescu) [inline src/Entity/EthereumServer.php]: As mentioned above, we shouldn't put these kinds of methods on the config entity, but in their specific controller instead.

File: src/Form/EthereumServerForm.php [hunk 11]
```diff
@@ -84,14 +84,12 @@ public function form(array $form, FormStateInterface $form_state) {
   public function save(array $form, FormStateInterface $form_state) {
     $server = $this->entity;
     $status = $server->save();
-
     if ($status == SAVED_UPDATED) {
-      $this->messenger()->addStatus($this->t('The server %label has been updated.', ['%label' => $server->label()]));
+      drupal_set_message($this->t('The server %label has been updated.', ['%label' => $server->label()]));
```
Comment (amateescu) [inline src/Form/EthereumServerForm.php]: `drupal_set_message()` is deprecated, the previous code is using the proper API.

File: src/Form/EthereumServerForm.php [hunk 12]
```diff
@@ -84,14 +84,12 @@ public function form(array $form, FormStateInterface $form_state) {
...
     }
     elseif ($status == SAVED_NEW) {
-      $this->messenger()->addStatus($this->t('The server %label has been added.', ['%label' => $server->label()]));
+      drupal_set_message($this->t('The server %label has been added.', ['%label' => $server->label()]));
```
Comment (amateescu) [inline src/Form/EthereumServerForm.php]: Same here.

File: src/Plugin/Field/FieldWidget/EthereumAddressWidget.php [hunk 13]
```diff
@@ -35,6 +35,10 @@ public static function defaultSettings() {
   public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
     $element = parent::formElement($items, $delta, $element, $form, $form_state);
 
+    if ($element['value']['#default_value']) {
+      $element['value']['#attributes']['data-original-value'] = $element['value']['#default_value'];
```
Comment (amateescu) [inline src/Plugin/Field/FieldWidget/EthereumAddressWidget.php:39]: Why do we need the original value here?

**Review by amateescu — CHANGES_REQUESTED** [review]
Reviewed the parts of the module that I'm familiar with :)

File: ethereum_user_connector/src/Plugin/rest/resource/ValidateEthereumAccount.php [hunk 14]
```diff
@@ -86,16 +86,19 @@ public static function create(ContainerInterface $container, array $configuratio
...
    *   With validated data.
    *   See ethereum_user_connector\Controller\verifyUserByHash().
    */
   public function get($hash) {
 
     if ($this->currentUser->isAuthenticated()) {
-      $eth = new EthereumUserConnectorController();
-      $validation = $eth->verifyUserByHash(Xss::filter($hash));
+
+      $controller = new EthereumUserConnectorController();
```
Comment (amateescu) [inline ethereum_user_connector/src/Plugin/rest/resource/ValidateEthereumAccount.php]: Here and in all the other places which instantiate `EthereumUserConnectorController`, we need to do it through the class resolver instead, like this: `\Drupal::service('class_resolver')->getInstanceFromDefinition(EthereumUserConnectorController::class)`

This will properly inject the dependencies from the container, and I'm pretty sure it's the reason for the changes to the constructor of `\Drupal\ethereum\Controller\EthereumController` (i.e. those changes can be reverted)

File: ethereum_txsigner/js/mascara/Readme.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
+# Metamask Mascara wrapper
```
Comment (amateescu) [inline ethereum_txsigner/js/mascara/Readme.md:1]: Shouldn't this be a stand-alone JS library (in a separate github repo) that the module would require?

File: ethereum_smartcontract/src/Entity/SmartContract.php (hunk 1 above)
Comment (digitaldonkey) [reply ethereum_smartcontract/src/Entity/SmartContract.php]: Moved to  ethereum_user_connector/src/Form/AdminForm.php

Comment (digitaldonkey) [reply ethereum_smartcontract/src/Entity/SmartContract.php]: Moved to form

File: ethereum_smartcontract/src/SmartContractInterface.php (hunk 2 above)
Comment (digitaldonkey) [reply ethereum_smartcontract/src/SmartContractInterface.php]: removed.

File: ethereum_txsigner/src/Entity/Txsigner.php (hunk 3 above)
Comment (digitaldonkey) [reply ethereum_txsigner/src/Entity/Txsigner.php:50]: removed

File: src/Controller/EthereumController.php (hunk 4 above)
Comment (digitaldonkey) [reply src/Controller/EthereumController.php]: Made protected and renamed to web3 for consistency.

File: src/Controller/EthereumController.php (hunk 6 above)
Comment (digitaldonkey) [reply src/Controller/EthereumController.php]: learning ... :heart:

File: src/Controller/EthereumController.php (hunk 7 above)
Comment (digitaldonkey) [reply src/Controller/EthereumController.php]: ok

File: src/Controller/EthereumController.php (hunk 8 above)
Comment (digitaldonkey) [reply src/Controller/EthereumController.php]: ok

File: src/Entity/EthereumServer.php (hunk 10 above)
Comment (digitaldonkey) [reply src/Entity/EthereumServer.php]: moved to form

File: src/Form/EthereumServerForm.php (hunk 11 above)
Comment (digitaldonkey) [reply src/Form/EthereumServerForm.php]: cleaned up all usages of drupal_set_message()

File: src/Form/EthereumServerForm.php (hunk 12 above)
Comment (digitaldonkey) [reply src/Form/EthereumServerForm.php]: cleaned up all usages of drupal_set_message()

File: src/Plugin/Field/FieldWidget/EthereumAddressWidget.php (hunk 13 above)
Comment (digitaldonkey) [reply src/Plugin/Field/FieldWidget/EthereumAddressWidget.php:39]: I use it to check if a address entry is the same than before, in order to give the address_status the right value. You are right: technically it should be somewhere in ethereum_user_connector module. There are 3 fields on the user required to manage the contract registry subscription (address, hash, address_status).

File: ethereum_txsigner/js/mascara/Readme.md (hunk 15 above)
Comment (digitaldonkey) [reply ethereum_txsigner/js/mascara/Readme.md:1]: Yes. It actually is a [external library](https://github.com/digitaldonkey/mascara_wrapper).

What I still miss is how to load external libs with composer. Seems we need to add asset-packagist to the root composer json like Thunder distribution does. Was hoping we will get that in core soon, before externalizing the dependencies.

File: src/Controller/EthereumController.php (hunk 5 above)
Comment (digitaldonkey) [reply src/Controller/EthereumController.php]: Hopefully fixed. This topic is still confusing for me.

File: src/Entity/EthereumServer.php (hunk 9 above)
Comment (digitaldonkey) [reply src/Entity/EthereumServer.php]: ok. Reverted to your code.
